storagenode/pieces/blobcache.go

```diff
@@ -0,0 +1,272 @@
+			service.log.Error("error during initializing space usage cache, saveCache: ", zap.Error(err))
+		}
+		return err
+	})
+}
+
+func (service *CacheService) persistCacheTotals(ctx context.Context) error {
+	cache := service.blobsUsageCache.cache
+	if err := service.store.spaceUsedDB.UpdateTotal(ctx, cache.total); err != nil {
+		return err
+	}
+	if err := service.store.spaceUsedDB.UpdateTotalsForAllSatellites(ctx, cache.totalBySatellite); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Init initializes the space used cache with the most recent values that were stored persistently
+func (service *CacheService) Init(ctx context.Context) (err error) {
+	total, err := service.store.spaceUsedDB.GetTotal(ctx)
+	if err != nil {
+		service.log.Error("CacheServiceInit error during initializing space usage cache GetTotal:", zap.Error(err))
+		return err
+	}
+	service.log.Debug("CacheServiceInit initializing cache GetTotal", zap.Int64("new total:", total))
+
+	totalBySatellites, err := service.store.spaceUsedDB.GetTotalsForAllSatellites(ctx)
+	if err != nil {
+		service.log.Error("CacheServiceInit error during initializing space usage cache GetTotalsForAllSatellites:", zap.Error(err))
+		return err
+	}
+	for _, saTotal := range totalBySatellites {
+		service.log.Debug("CacheServiceInit initializing cache SA total", zap.Int64("new totalBySA:", saTotal))
+	}
+
+	service.blobsUsageCache.init(total, totalBySatellites)
+	return nil
+}
+
+// Close closes the loop
+func (service *CacheService) Close() (err error) {
+	service.loop.Close()
+	return nil
+}
+
+func (blobs *BlobsUsageCache) init(total int64, totalBySatellite map[storj.NodeID]int64) {
+	blobs.mu.Lock()
+	defer blobs.mu.Unlock()
+	blobs.cache.total = total
+	blobs.cache.totalBySatellite = totalBySatellite
+}
+
+// SpaceUsedBySatellite returns the current total space used for a specific
+// satellite for all pieces (not including header bytes)
+func (blobs *BlobsUsageCache) SpaceUsedBySatellite(ctx context.Context, satelliteID storj.NodeID) (int64, error) {
+	blobs.mu.Lock()
+	defer blobs.mu.Unlock()
+	return blobs.cache.totalBySatellite[satelliteID], nil
+}
+
+// SpaceUsedForPieces returns the current total used space for
+//// all pieces content (not including header bytes)
+func (blobs *BlobsUsageCache) SpaceUsedForPieces(ctx context.Context) (int64, error) {
+	blobs.mu.Lock()
+	defer blobs.mu.Unlock()
+	return blobs.cache.total, nil
+}
+
+func (blobs *BlobsUsageCache) update(ctx context.Context, satelliteID storj.NodeID, pieceContentSize int64) {
+	blobs.mu.Lock()
+	defer blobs.mu.Unlock()
+	blobs.cache.total += pieceContentSize
+	blobs.cache.totalBySatellite[satelliteID] += pieceContentSize
+}
+
+// Close satisfies the pieces interface
+func (blobs *BlobsUsageCache) Close() error {
+	return nil
+}
+
+func (blobs *BlobsUsageCache) recalculate(ctx context.Context, newTotalSpaceUsed int64, newTotalSpaceUsedByNamespace map[storj.NodeID]int64) error {
+	spaceUsedWhenIterationStarted := blobs.cache
```

Comment:
isn't there a small data race here where an update call can come in while recalculate is happening? and by small i mean "formally incorrect so possibly triggered by the race detector but would never actually happen in practice."

also, this isn't really the space used when the iteration started, this is also the space used when the iteration ended. the iteration happens before we even call into recalculate.

Reply:
yeah this variable name is wrong, it should be "Ended" not "Started". Also I need to fix the code, because I'm not even tracking the "Started" case. Oops. 

About the data race situation... currently we recalculate and reset the in-memory used space values on startup, but while that is occurring, some updates could come in. I asked an open question about that and here is a possible solution that Paul came up with https://github.com/storj/storj/pull/2707#issuecomment-517965671.  I decided to implement that here, but I actually did it wrong, need to fix that now.

Reply:
That comment/calculation is useful to handle the **race condition**, but I’m talking about a **data race**: two concurrent processes accessing the same memory location where at least one is a write. In other words, the update call protects it’s write with a mutex, and since an update can be concurrent with this read, the read must also be protected by a mutex.

Reply:
i can add a mu.lock here as well, i just assumed since the result of `recalculate` is an estimation anyways that the race condition here didn't actually matter.

Reply:
i think it’s always important to have no data races, even benign ones. this way, there’s no false positives with the race detector, and there are good arguments that there’s no such thing as a benign data race since optimizing compilers assume your program is free of them. another great article: https://software.intel.com/en-us/blogs/2013/01/06/benign-data-races-what-could-possibly-go-wrong

Reply:
that all makes sense. and yeah seems better to avoid race conditions all together. thanks. i will fix.

Reply:
and wow that intel blog really drives home the point. ok ok i will add a lock. haha